tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java

```diff
@@ -221,43 +379,86 @@ public void newWindow() {
-            double recsPerSec = 1000.0 * count / (double) elapsed;
-            double mbPerSec = 1000.0 * this.bytes / (double) elapsed / (1024.0 * 1024.0);
-            int[] percs = percentiles(this.latencies, index, 0.5, 0.95, 0.99, 0.999);
-            System.out.printf("%d records sent, %f records/sec (%.2f MB/sec), %.2f ms avg latency, %.2f ms max latency, %d ms 50th, %d ms 95th, %d ms 99th, %d ms 99.9th.\n",
-                              count,
-                              recsPerSec,
-                              mbPerSec,
-                              totalLatency / (double) count,
-                              (double) maxLatency,
-                              percs[0],
-                              percs[1],
-                              percs[2],
-                              percs[3]);
+    private static final class ProducerPerformanceThread extends Thread {
+        private final Producer producer;
+        private final String topicName;
+        private final long numRecords;
+        private final int recordSize;
+        private final int throughput;
+        private final int valueBound;
+        private final MetricStats metricStats;
+        private final ThreadStats threadStats;
+        private final Random random = new Random();
+
+        ProducerPerformanceThread(Producer producer,
+                                  String topicName,
+                                  long numRecords,
+                                  int recordSize,
+                                  int throughput,
+                                  int valueBound,
+                                  MetricStats metricStats,
+                                  ThreadStats threadStats) {
+            this.producer = producer;
+            this.topicName = topicName;
+            this.numRecords = numRecords;
+            this.recordSize = recordSize;
+            this.throughput = throughput;
+            this.valueBound = valueBound;
+            this.metricStats = metricStats;
+            this.threadStats = threadStats;
         }
 
-        private static int[] percentiles(int[] latencies, int count, double... percentiles) {
-            int size = Math.min(count, latencies.length);
-            Arrays.sort(latencies, 0, size);
-            int[] values = new int[percentiles.length];
-            for (int i = 0; i < percentiles.length; i++) {
-                int index = (int) (percentiles[i] * size);
-                values[i] = latencies[index];
+        @Override
+        public void run() {
+
+            long startMs = System.currentTimeMillis();
+            ThroughputThrottler throttler = new ThroughputThrottler(throughput, startMs);
+            int messagesSinceLastMetricRecord = 0;
+            int metricStatsInterval = (int) numRecords / 20;
+            int skipMessagesForMetricSampling = (int) numRecords / 2;
+            for (int i = 0; i < numRecords; i++) {
+                byte[] payload = newPayload();
```

Comment:
@ewencp I was also wondering if this would impact the performance so I ran some tests using preallocated payload. I did not see statistical difference last time when I ran the tests (<2% if any).

I actually used to use preallocated payloads, but in reality it might be a little difficult to decide how big N should be because it also depends on how many partitions the topic has and how big the batch is. In worst case, we may end up using a lot of memory to store the preallocated payload.

Dynamically generating the payload might be OK in our case because now we support multiple user threads, so if generating payload becomes the bottleneck we can always increase the number of user threads.